### ISSUE TITLE
Not possible to retrieve the hostname of a replica (swarm) with Docker 1.12.5

### ISSUE THREAD

**Issue by GabyCT** [01org/cc-oci-runtime#578]
Currently in docker 1.12.5 it is not possible to obtain the hostname of any replicas, as curl hangs all the time. However, this behavior is already fixed in Docker version 1.13.0-rc5 using runc.

1.  docker swarm init
2. docker service create --name testswarm --replicas 4 --publish 8080:80 nginx /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \daemon off;\" 2> /dev/null
3. unset http_proxy
4. curl http://127.0.0.1:8080/hostname 
(here it hangs)

**Comment by GabyCT** [comment]
Using runc with docker 1.12.5 is working perfectly.
1. url=http://127.0.0.1:8080/hostname
2. curl $url
7ca4aa689c16

This was tested on using clearlinux.

**Comment by GabyCT** [comment]
Swarm should be tested using 2.1

**Comment by devimc** [comment]
Port forwarding using ```docker --publish``` works
```
$ docker run --publish 8080:80 -dti nginx bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\""
4fc876a718329d74ff066a8dba895f7558218cfe5a206265f91b972766af0974
$ unset http_proxy
$ curl http://127.0.0.1:8080/hostname
4fc876a71832
```
Port forwarding with swarm is NOT working using ```http://127.0.0.1:8080/``` but it works using container's IP
```
$ docker swarm init --advertise-addr enp0s25
$ docker service create --name testswarm --replicas 1 --publish 8080:80 nginx /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\"" 2> /dev/null
$ docker exec af9baed6d0ac ip a
...
3: eth1: <BROADCAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 02:42:ac:12:00:03 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:3/64 scope link 
       valid_lft forever preferred_lft forever
....
$ curl http://172.18.0.3/hostname
af9baed6d0ac
```
notice that the IP of this container is **172.18.0.3**
I executed ```iptables``` to get more information about port forwarding
```
# iptables -t nat -vnL | grep 8080
    3   180 DNAT       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8080 to:172.18.0.2:8080
```
```curl http://172.18.0.2:8080/hostname``` hangs in cc-oci-runtime but it doesn't hang in runc it returns the hostname of 172.18.0.3

@mcastelino do you have any idea?

**Comment by devimc** [comment]
debugging ```curl http://localhost:8080/hostname``` using tcpdump

**cc-oci-runtime**
```
# tcpdump -vv -i docker_gwbridge 
tcpdump: listening on docker_gwbridge, link-type EN10MB (Ethernet), capture size 262144 bytes
09:51:07.547321 IP (tos 0x0, ttl 64, id 46725, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0xd589), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4051905 ecr 0,nop,wscale 7], length 0
09:51:08.569459 IP (tos 0x0, ttl 64, id 46726, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0xd18a), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4052928 ecr 0,nop,wscale 7], length 0
09:51:10.617425 IP (tos 0x0, ttl 64, id 46727, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0xc98a), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4054976 ecr 0,nop,wscale 7], length 0
09:51:14.649414 IP (tos 0x0, ttl 64, id 46728, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0xb9ca), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4059008 ecr 0,nop,wscale 7], length 0
09:51:22.969472 IP (tos 0x0, ttl 64, id 46729, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0x994a), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4067328 ecr 0,nop,wscale 7], length 0
09:51:28.089443 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 172.18.0.2 tell 172.18.0.1, length 28
09:51:28.089491 ARP, Ethernet (len 6), IPv4 (len 4), Reply 172.18.0.2 is-at 02:42:ac:12:00:02 (oui Unknown), length 28
09:51:39.353460 IP (tos 0x0, ttl 64, id 46730, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35270 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0x594a), seq 2464572740, win 43690, options [mss 65495,sackOK,TS val 4083712 ecr 0,nop,wscale 7], length 0
```

**runC**
```
# tcpdump -vv -i docker_gwbridge 
tcpdump: listening on docker_gwbridge, link-type EN10MB (Ethernet), capture size 262144 bytes
09:54:10.410177 IP (tos 0x0, ttl 64, id 22719, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.1.35568 > 172.18.0.2.webcache: Flags [S], cksum 0x5856 (incorrect -> 0x55e5), seq 3852715695, win 43690, options [mss 65495,sackOK,TS val 4234768 ecr 0,nop,wscale 7], length 0
09:54:10.410248 IP (tos 0x0, ttl 63, id 0, offset 0, flags [DF], proto TCP (6), length 60)
	172.18.0.2.webcache > localhost.35568: Flags [S.], cksum 0x2b44 (incorrect -> 0xc8c1), seq 2840046164, ack 3852715696, win 27960, options [mss 1410,sackOK,TS val 4234768 ecr 4234768,nop,wscale 7], length 0
09:54:10.410266 IP (tos 0x0, ttl 64, id 22720, offset 0, flags [DF], proto TCP (6), length 52)
	172.18.0.1.35568 > 172.18.0.2.webcache: Flags [.], cksum 0x584e (incorrect -> 0x362c), seq 3852715696, ack 2840046165, win 342, options [nop,nop,TS val 4234768 ecr 4234768], length 0
09:54:10.410313 IP (tos 0x0, ttl 64, id 22721, offset 0, flags [DF], proto TCP (6), length 138)
	172.18.0.1.35568 > 172.18.0.2.webcache: Flags [P.], cksum 0x58a4 (incorrect -> 0xe6c8), seq 0:86, ack 1, win 342, options [nop,nop,TS val 4234768 ecr 4234768], length 86: HTTP, length: 86
	GET /hostname HTTP/1.1
	Host: localhost:8080
	User-Agent: curl/7.43.0
	Accept: */*
	
09:54:10.410333 IP (tos 0x0, ttl 63, id 13381, offset 0, flags [DF], proto TCP (6), length 52)
	172.18.0.2.webcache > localhost.35568: Flags [.], cksum 0x2b3c (incorrect -> 0x6363), seq 1, ack 87, win 219, options [nop,nop,TS val 4234768 ecr 4234768], length 0
09:54:10.410402 IP (tos 0x0, ttl 63, id 13382, offset 0, flags [DF], proto TCP (6), length 302)
	172.18.0.2.webcache > localhost.35568: Flags [P.], cksum 0x2c36 (incorrect -> 0xac6a), seq 1:251, ack 87, win 219, options [nop,nop,TS val 4234769 ecr 4234768], length 250: HTTP, length: 250
	HTTP/1.1 200 OK
	Server: nginx/1.11.8
	Date: Thu, 19 Jan 2017 15:54:10 GMT
	Content-Type: application/octet-stream
	Content-Length: 13
	Last-Modified: Thu, 19 Jan 2017 15:53:51 GMT
	Connection: keep-alive
	ETag: "5880e10f-d"
	Accept-Ranges: bytes
	
09:54:10.410417 IP (tos 0x0, ttl 64, id 22722, offset 0, flags [DF], proto TCP (6), length 52)
	172.18.0.1.35568 > 172.18.0.2.webcache: Flags [.], cksum 0x584e (incorrect -> 0x34d2), seq 86, ack 251, win 350, options [nop,nop,TS val 4234769 ecr 4234769], length 0
```

one thing to highlight is that **cc-oci-runtime** has
```
172.18.0.1.35270 > 172.18.0.2.webcache
```
but  **runc** has
```
172.18.0.1.35568 > 172.18.0.2.webcache
172.18.0.2.webcache > localhost.35568
```

**Comment by devimc** [comment]
debugging 
```
$ docker swarm init --advertise-addr enp0s25
$ docker service create --name testswarm --replicas 1 --publish 8080:80 nginx /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\""
```

**cc-oci-runtime**
```
# tcpdump -vv -i docker_gwbridge 
tcpdump: listening on docker_gwbridge, link-type EN10MB (Ethernet), capture size 262144 bytes
10:29:28.897395 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:2 to_ex { }]
10:29:29.881460 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:2 to_ex { }]
10:29:29.881530 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 24) :: > ff02::1:ff12:2: [icmp6 sum ok] ICMP6, neighbor solicitation, length 24, who has fe80::42:acff:fe12:2
10:29:30.905498 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:2 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:2 to_ex { }]
10:29:30.905541 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:2 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:02
		0x0000:  0242 ac12 0002
10:29:31.081461 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:2 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:2 to_ex { }]
10:29:34.937452 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:2 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:02
		0x0000:  0242 ac12 0002
10:29:35.206469 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:38ff:fec7:c68f > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::fb to_ex { }]
10:29:35.625435 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:38ff:fec7:c68f > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::fb to_ex { }]
10:29:38.969388 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:2 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:02
		0x0000:  0242 ac12 0002
10:29:39.999405 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:40.008378 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:fffe:1 to_ex { }]
10:29:40.260903 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 172.18.0.1 tell 172.18.0.3, length 28
10:29:40.260934 ARP, Ethernet (len 6), IPv4 (len 4), Reply 172.18.0.1 is-at 02:42:38:c7:c6:8f (oui Unknown), length 28
10:29:40.261003 IP (tos 0x10, ttl 64, id 31290, offset 0, flags [DF], proto UDP (17), length 76)
	172.18.0.3.46784 > 172.18.0.1.ntp: [bad udp cksum 0x5872 -> 0x5014!] NTPv4, length 48
	Client, Leap indicator:  (0), Stratum 0 (unspecified), poll 0 (1s), precision 0
	Root Delay: 0.000000, Root dispersion: 0.000000, Reference-ID: (unspec)
	  Reference Timestamp:  0.000000000
	  Originator Timestamp: 0.000000000
	  Receive Timestamp:    0.000000000
	  Transmit Timestamp:   3693832180.059816930 (2017/01/19 10:29:40)
		Originator - Receive Timestamp:  0.000000000
		Originator - Transmit Timestamp: 3693832180.059816930 (2017/01/19 10:29:40)
10:29:40.261047 IP (tos 0xd0, ttl 64, id 4870, offset 0, flags [none], proto ICMP (1), length 104)
	172.18.0.1 > 172.18.0.3: ICMP host 172.18.0.1 unreachable - admin prohibited, length 84
	IP (tos 0x10, ttl 64, id 31290, offset 0, flags [DF], proto UDP (17), length 76)
	172.18.0.3.46784 > 172.18.0.1.ntp: [bad udp cksum 0x5872 -> 0x5014!] NTPv4, length 48
	Client, Leap indicator:  (0), Stratum 0 (unspecified), poll 0 (1s), precision 0
	Root Delay: 0.000000, Root dispersion: 0.000000, Reference-ID: (unspec)
	  Reference Timestamp:  0.000000000
	  Originator Timestamp: 0.000000000
	  Receive Timestamp:    0.000000000
	  Transmit Timestamp:   3693832180.059816930 (2017/01/19 10:29:40)
		Originator - Receive Timestamp:  0.000000000
		Originator - Transmit Timestamp: 3693832180.059816930 (2017/01/19 10:29:40)
10:29:40.261260 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:40.569414 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 24) :: > ff02::1:fffe:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 24, who has fe80::caff:fefe:1
10:29:40.762280 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:40.825429 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 24) :: > ff02::1:ff12:3: [icmp6 sum ok] ICMP6, neighbor solicitation, length 24, who has fe80::42:acff:fe12:3
10:29:40.889404 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:40.942410 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 24) :: > ff02::1:ff12:3: [icmp6 sum ok] ICMP6, neighbor solicitation, length 24, who has fe80::42:acff:fe12:3
10:29:40.953437 IP6 (hlim 1, next-header Options (0) payload length: 36) :: > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:fffe:1 to_ex { }]
10:29:41.593505 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::caff:fefe:1 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:fffe:1 to_ex { }]
10:29:41.593563 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::caff:fefe:1 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:41.833456 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::caff:fefe:1 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:fffe:1 to_ex { }]
10:29:41.849490 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:3 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:41.849543 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:41.944413 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:3 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:41.944467 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:03
		0x0000:  0242 ac12 0003
10:29:42.094361 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:3 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:42.681449 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:acff:fe12:3 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::1:ff12:3 to_ex { }]
10:29:45.433385 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 172.18.0.3 tell 172.18.0.1, length 28
10:29:45.433570 ARP, Ethernet (len 6), IPv4 (len 4), Reply 172.18.0.3 is-at 02:42:ac:12:00:03 (oui Unknown), length 28
10:29:45.625454 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::caff:fefe:1 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:45.881449 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:45.952273 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:03
		0x0000:  0242 ac12 0003
10:29:46.214472 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:38ff:fec7:c68f > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::fb to_ex { }]
10:29:46.361406 IP6 (hlim 1, next-header Options (0) payload length: 36) fe80::42:38ff:fec7:c68f > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 1 group record(s) [gaddr ff02::fb to_ex { }]
10:29:49.657452 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::caff:fefe:1 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:49.913428 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:00:ca:fe:00:01
		0x0000:  0200 cafe 0001
10:29:49.960295 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 16) fe80::42:acff:fe12:3 > ff02::2: [icmp6 sum ok] ICMP6, router solicitation, length 16
	  source link-address option (1), length 8 (1): 02:42:ac:12:00:03
		0x0000:  0242 ac12 0003
```

one thing to highlight
```
172.18.0.1 > 172.18.0.3: ICMP host 172.18.0.1 unreachable - admin prohibited, length 84
```

investigating firewall & iptables ....

**Comment by mcastelino** [comment]
@devimc @jcvenegas Looking at this now. I am able to replicate it.

**Comment by mcastelino** [comment]
Root caused this issue. There were 3 bugs in our code

1. MTU is not being setup on the interfaces (on the tap interface as well as the interface inside the VM)
2. Docker sets up port mapping rules inside the container ns (so we need to propagate them into the VM)
3. The container side iptables rules *seem* to interfere with the flow of bridged packets into the VM. This is probably bridge filtering pushing traffic to the IP table rules.

The first two need fixes in our code. Both in the hyperstart agent inside the VM as well as the runtime.

I am investigating the root cause of 3.

More details

**Configuration:**
Container Ingress IP: 10.255.0.4
Container Ingress VIP:10.255.0.2

Gateway Bridge IP: 172.18.0.1

Ingress Sbox Ingress IP: 10.255.0.3 (talks to Container Ingress IP)
Ingress Sbox gateway IP: 172.18.0.2 (talks to sbox gatewat network)

So traffic flow localhost -> docker_gwbridge_172.18.0.1 -> ingress_sbox_172.18.0.2-> marked traffic ip_vs -10.255.0.2----> container_10.255.0.4

**Host iptables that matter:**
How the traffic gets in/out of the ingress_sbox
```
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER-INGRESS
-A DOCKER-INGRESS -p tcp -m tcp --dport 8080 -j DNAT --to-destination 172.18.0.2:8080
-A DOCKER-INGRESS -p tcp -m state --state RELATED,ESTABLISHED -m tcp --sport 8080 -j ACCEPT

```
Ingress sbox rules: nsenter --net=/var/run/docker/netns/ingress_sbox
```
-A PREROUTING -p tcp -m tcp --dport 8080 -j MARK --set-xmark 0x100/0xffffffff
-A POSTROUTING -d 10.255.0.0/16 -m ipvs --ipvs -j SNAT --to-source 10.255.0.3
```

IPVS Setup in ingress_sbox - This sends traffic coming in to 10.255.0.4
```
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
FWM  256 rr
  -> 10.255.0.4:0                 Masq    1      0          0

ipvsadm-save
-A -f 256 -s rr
-a -f 256 -r 10.255.0.4:0 -m -w 1
```

Note: 0x100 == 256 :)

**Connecting to the container from the ingress_sbox (both 80 and 8080 work????)**
root@singlevm:/var/run/docker/netns# curl http://10.255.0.4:80/hostname
1f99d9cf0236
root@singlevm:/var/run/docker/netns# curl http://10.255.0.4:8080/hostname
1f99d9cf0236

**However in the case of the cors container the published port does not work**
root@singlevm:/home/mrcastel# curl http://10.255.0.5:80/hostname
b578e43a836d
root@singlevm:/home/mrcastel# curl http://10.255.0.5:8080/hostname
^C

Also traffic comes into the ingress sbox but never gets out????
**Docker setup up iptables rules INSIDE THE CONTAINER NS**
Get into the ns of the VM: not inside the VM
nsenter --net=/var/run/docker/netns/02d51fa13d84

```
-A PREROUTING -d 10.255.0.4/32 -p tcp -m tcp --dport 8080 -j REDIRECT --to-ports 80
-A OUTPUT -d 127.0.0.11/32 -j DOCKER_OUTPUT
-A POSTROUTING -d 127.0.0.11/32 -j DOCKER_POSTROUTING
-A DOCKER_OUTPUT -d 127.0.0.11/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 127.0.0.11:41343
-A DOCKER_OUTPUT -d 127.0.0.11/32 -p udp -m udp --dport 53 -j DNAT --to-destination 127.0.0.11:43411
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p tcp -m tcp --sport 41343 -j SNAT --to-source :53
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p udp -m udp --sport 43411 -j SNAT --to-source :53
COMMIT
# Completed on Thu Feb  2 21:57:42 2017
# Generated by iptables-save v1.6.0 on Thu Feb  2 21:57:42 2017
*filter
:INPUT ACCEPT [67:5781]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [75:5765]
-A INPUT -d 10.255.0.4/32 -p tcp -m tcp --dport 80 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
-A INPUT -d 10.255.0.4/32 -p udp -j DROP
-A INPUT -d 10.255.0.4/32 -p tcp -j DROP
-A OUTPUT -s 10.255.0.4/32 -p tcp -m tcp --sport 80 -m conntrack --ctstate ESTABLISHED -j ACCEPT
-A OUTPUT -s 10.255.0.4/32 -p udp -j DROP
-A OUTPUT -s 10.255.0.4/32 -p tcp -j DROP

```

To test this hypothesis I created a new nginx image with port 8080 and now the iptables port remapping should not matter

docker service create --name testswarm --replicas 1 --publish 8080:8080 mcastelino/nginx8080 /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\"" 2> /dev/null

**Even with this the traffic did not get through to the VM. It reached the veth interface and did not get out the bridge.**

I had to clear the iptable rules in the VM NS to actually get the request all the way through.

So the last step is to figure out why the host side ns rules interfere with purely bridges traffic.
Currently I assume it is due to bridge filtering issues.

**Comment by sameo** [comment]
> 1. MTU is not being setup on the interfaces (on the tap interface as well as the interface inside the VM)

We should have an open issue for that one. @amshinde If you're working on it, please open an issue to track it.

**Comment by amshinde** [comment]
@sameo Sure. I'll open issues for setting up port-mapping as well.

**Comment by GabyCT** [comment]
closing this issue as this was already solved with PRs #831 and #825 and the tests are already unskipped  (047fd21).
